core/trino-main/src/main/java/io/trino/operator/aggregation/DistinctAccumulatorFactory.java

```diff
@@ -114,7 +117,12 @@ private DistinctAccumulator(
                     Optional.empty(),
                     joinCompiler,
                     blockTypeOperators,
-                    UpdateMemory.NOOP);
+                    () -> {
+                        // enforce task memory limits for fast throw
+                        updateMemory.update();
+                        // never block, as addInput doesn't support yield semantics
+                        return true;
```

Comment:
This seems broken.  `UpdateMemory` never throws (it isn't allowed), and you can still get an OOM.  If we are going to add this, yield should be added to addInput.

Reply:
Do you actually need to use `boolean update();` method or `void update();` would be sufficient. If so, I suggest to have an interface with the latter, so it's not confusing

Reply:
@sopel39 dug a lot deeper and found that this does throw today when user memory is exceeded.  In that case, this code seems totally reasonable.

Reply:
> Do you actually need to use `boolean update();` method or `void update();` would be sufficient. If so, I suggest to have an interface with the latter, so it's not confusing
> 
> Maybe you could:
> 
> * introduce a class `NonYieldingUpdateMemory` or
> * you could rename current `UpdateMemory` to `YieldingUpdateMemory` and have `void update()` in `UpdateMemory`

I'm not sure that is a better way. Even if we have 'void update();', 'return true;' cannot be deleted since MarkDistinctHash requires 'boolean update();'. So, it will be just changing the type. Also, 'from where to use NonYieldingUpdateMemory' could be ambiguous(from AggregatorFactory.createAggregator or from DistinctAccumulator's constructor or etc).